torch/_functorch/functional_assertions_helper.py

```diff
@@ -0,0 +1,103 @@
+from collections import defaultdict
+from typing import Callable, Dict
+
+import torch
+import torch._decomp as decomp
+from torch._ops import OpOverload
+import torch._functorch.config as config
+
+aten = torch.ops.aten
+_decompositions: Dict[str, Dict[OpOverload, Callable]] = defaultdict(dict)
+
+
+class DepTokenStateTracker:
+    """
+    A singleton / global context manager which is used to:
+    - Record the global flag that assertion ops functionalization is enabled
+      during compilation if the context is entered.
+    - Record most recent token which is propagated through functional assertion
+      ops.
+    """
+
+    _dep_token = None
+    _entered: bool = False
+
+    def __enter__(self) -> "DepTokenStateTracker":
+        assert (
+            not DepTokenStateTracker._entered
+        ), "Cannot reenter singleton context `DepTokenStateTracker`."
+        self.reset()
+        DepTokenStateTracker._entered = True
+        return self
+
+    def __exit__(self, exc_type, exc_cal, exc_tb) -> None:
+        assert DepTokenStateTracker._entered, (
+            "Cannot exit singleton context `DepTokenStateTracker` which is not "
+            "entered previously."
+        )
+        self.reset()
+        DepTokenStateTracker._entered = False
+        return None
+
+    @classmethod
+    def reset(cls) -> None:
+        cls._dep_token = None
+
+    @classmethod
+    def set_dep_token(cls, dep_token) -> None:
+        assert cls._entered
+        cls._dep_token = dep_token
+
+    @classmethod
+    def get_dep_token(cls):
+        if cls._dep_token is None:
+            cls._dep_token = aten._make_dep_token()
+
+        return cls._dep_token
+
+
+class FunctionalAssertionsHelper:
+    @staticmethod
+    def can_functionalize(needs_autograd: bool) -> bool:
+        if config.functionalize_assertion_ops:
```

Comment:
one question: will we ever **not** want to functionalize asserts for export? I'm wondering if this config is meant to be a long-term option (so users can decide to or not to functionalize asserts during export), or if we'll eventually want it to always be enabled.

If we want it to be always-on but only for export, one option would be to check the flag `aot_config.is_export()`, that tells us if we're running from an `aot_export_module()` call, vs the vanilla torch.compile() codepaths.

Reply:
@tugsbayasgalan: what's your thought here? I feel like maybe for now we still want to provide a way for user to opt out assertion functionalization just for safety?

> In fact, I think we probably want this check, so nobody accidentally tries to functionalize asserts while running `torch.compile`

So you're saying for eager mode (non-export) we want to make assertion functionazation opt in based (by default no functionalization)?

Reply:
Yeah I think we still need to provide a way to opt out of functionalization assertion as the downstream users are still modifying the stack to adopt our new changes to the graph. So I feel little uneasy about turning it on unconditionally yet.